source/app/libpd.c

```diff
@@ -216,6 +251,128 @@ void parodus_receive_wait()
+int getConnCloudStatus(char *device_mac)
+{
+	char *source = NULL, *dest = NULL;
+	wrp_msg_t *req_wrp_msg = NULL;
+	char *contentType = NULL;
+	int rv = -1;
+	int sendStatus = -1;
+    int backoffRetryTime = 0;
+    int c=2;
+    char *transaction_uuid = NULL;
+
+	if(device_mac == NULL)
+	{
+		WalError("device_mac is NULL, unable to get cloud_status\n");
+		return rv;
+	}
+	else
+	{
+		req_wrp_msg = (wrp_msg_t *)malloc(sizeof(wrp_msg_t));
+		if(req_wrp_msg != NULL)
+		{
+			memset(req_wrp_msg, 0, sizeof(wrp_msg_t));
+			req_wrp_msg->msg_type = WRP_MSG_TYPE__RETREIVE;
+
+			source = (char *) malloc(sizeof(char)*MAX_STR_LENGTH);
+			dest   = (char *) malloc(sizeof(char)*MAX_STR_LENGTH);
+
+			if(source !=NULL)
+			{
+				snprintf(source, MAX_STR_LENGTH, "mac:%s/config", device_mac);
+				req_wrp_msg->u.crud.source = source;
+				WalPrint("req_wrp_msg->u.crud.source is %s\n", req_wrp_msg->u.crud.source);
+			}
+
+			if(dest !=NULL)
+			{
+				snprintf(dest, MAX_STR_LENGTH, "mac:%s/parodus/cloud-status", device_mac);
+				req_wrp_msg->u.crud.dest = dest;
+				WalPrint("req_wrp_msg->u.crud.dest is %s\n", req_wrp_msg->u.crud.dest);
+			}
+
+			generate_trans_uuid(&transaction_uuid);
+			if(transaction_uuid !=NULL)
+			{
+				req_wrp_msg->u.crud.transaction_uuid = transaction_uuid;
+				WalInfo("transaction_uuid generated is %s\n", req_wrp_msg->u.crud.transaction_uuid);
+			}
+
+			contentType = strdup(CONTENT_TYPE_JSON);
+			if(contentType != NULL)
+			{
+				req_wrp_msg->u.crud.content_type = contentType;
+				WalPrint("retrieve content_type is %s\n",req_wrp_msg->u.crud.content_type);
+			}
+
+			while( FOREVER() )
+			{
+				backoffRetryTime = (int) pow(2, c) -1;
+				WalPrint("Backoff calculated is %d\n", backoffRetryTime);
+
+				sendStatus = libparodus_send(current_instance, req_wrp_msg);
+				WalPrint("sendStatus is %d\n",sendStatus);
+				if(sendStatus == 0)
+				{
+					WalInfo("Sent retrieve request successfully to parodus\n");
+				}
+				else
+				{
+					WalError("Failed to send retrieve req: '%s'\n",libparodus_strerror(sendStatus));
+					break;
+				}
+
+				//waiting to get response from parodus. add lock here while reading
+				pthread_mutex_lock (&cloud_mut);
+				WalPrint("mutex lock in consumer thread\n");
+				WalPrint("Before pthread cond wait in consumer thread\n");
+				pthread_cond_wait(&cloud_con, &cloud_mut);
+				pthread_mutex_unlock (&cloud_mut);
+				WalPrint("mutex unlock in consumer thread after cond wait\n");
+
+				if ((get_global_cloud_status() !=NULL) && (strcmp(get_global_cloud_status(), CLOUD_STATUS_ONLINE) == 0))
+				{
+					WalPrint("Received cloud_status as online, returning ..\n");
+					rv = 1;
+					free(get_global_cloud_status());
+					break;
+				}
+				else
+				{
+					WalError("cloud_status is not online, sending RETRIEVE request again and retrying\n" );
+					WalInfo("cloud status retry backoffRetryTime '%d' seconds\n", backoffRetryTime);
+					sleep(backoffRetryTime);
+					c++;
```

Comment:
c here in backoffRetryTime should be maxed to 9 i.e. ~511s and not keep incrementing